perception/autoware_ground_segmentation/src/scan_ground_filter/grid_ground_filter.cpp

```diff
@@ -0,0 +1,460 @@
+    if (delta_avg_z > param_.detection_range_z_max) {
+      // this point is out-of-range
+      continue;
+    }
+
+    // 2. the angle is exceed the global slope threshold
+    if (height > param_.global_slope_max_ratio * radius) {
+      // this point is obstacle
+      out_no_ground_indices.indices.push_back(pt_idx);
+      // go to the next point
+      continue;
+    }
+    // 3. local slope
+    const float delta_radius = radius - prev_cell.avg_radius_;
+    const float global_slope_threshold = param_.global_slope_max_ratio * delta_radius;
+    if (abs(delta_avg_z) < global_slope_threshold) {
+      // this point is ground
+      ground_bin.addPoint(radius, height, pt_idx);
+      // go to the next point
+      continue;
+    }
+    // 4. height from the estimated ground
+    if (abs(delta_avg_z) < param_.non_ground_height_threshold) {
+      // this point is ground
+      ground_bin.addPoint(radius, height, pt_idx);
+      // go to the next point
+      continue;
+    }
+    const float delta_max_z = height - prev_cell.max_height_;
+    if (abs(delta_max_z) < param_.non_ground_height_threshold) {
+      // this point is ground
+      ground_bin.addPoint(radius, height, pt_idx);
+      // go to the next point
+      continue;
+    }
+    // 5. obstacle from local slope
+    if (delta_avg_z >= global_slope_threshold) {
+      // this point is obstacle
+      out_no_ground_indices.indices.push_back(pt_idx);
+      // go to the next point
+      continue;
+    }
+  }
+}
+
+void GridGroundFilter::SegmentBreakCell(
+  const Cell & cell, PointsCentroid & ground_bin, pcl::PointIndices & out_no_ground_indices)
+{
+  const Cell & prev_cell = grid_ptr_->getCell(cell.scan_grid_root_idx_);
+
+  // loop over points in the cell
+  for (const auto & pt : cell.point_list_) {
+    const size_t & pt_idx = pt.index;
+    const float & radius = pt.distance;
+    const float & height = pt.height;
+
+    // 1. height is out-of-range
+    const float delta_z = height - prev_cell.avg_height_;
+    if (delta_z > param_.detection_range_z_max) {
+      // this point is out-of-range
+      continue;
+    }
+
+    // 2. the angle is exceed the global slope threshold
+    if (height > param_.global_slope_max_ratio * radius) {
+      // this point is obstacle
+      out_no_ground_indices.indices.push_back(pt_idx);
+      // go to the next point
+      continue;
+    }
+
+    // 3. the point is over discontinuous grid
+    const float delta_radius = radius - prev_cell.avg_radius_;
+    const float global_slope_threshold = param_.global_slope_max_ratio * delta_radius;
+    if (abs(delta_z) < global_slope_threshold) {
+      // this point is ground
+      ground_bin.addPoint(radius, height, pt_idx);
+      // go to the next point
+      continue;
+    }
+    if (delta_z >= global_slope_threshold) {
+      // this point is obstacle
+      out_no_ground_indices.indices.push_back(pt_idx);
+      // go to the next point
+      continue;
+    }
+  }
```

Comment:
[imo]

Beware underground pointclouds won't be segmented to ground nor obstacle.